doc: Add NixOS/NUC handbook (#2631)
m-labs/artiq: doc/manual/nuc_handbook.rst

NixOS for ARTIQ (NUC handbook)
==============================

> ARTIQ is a fully open-source project, and can be installed and used on any operating system. All else being equal, however, M-Labs suggests using the Linux distribution `NixOS <https://nixos.org/>`_, based around the Nix package manager, which ARTIQ is also distributed through. NixOS is used in-house by M-Labs to build, test and develop ARTIQ; if you have total freedom of OS and you want to quickly set up an ARTIQ environment that will *definitely work*, NixOS should be your first choice. Even if you are familiar with Linux, it has some peculiarities, but the advantages it offers (reproduceability, reliability, stability) are worth the learning curve.

Reviewer: > ARTIQ [...] can be installed and used on any operating system.

As much as everyone appreciates good documentation, that's obscenely wrong.
The entire line of argument appears pointless and presumptuous for a "handbook". After all, the explicit assumption is that the intended audience already made the choice.
And what does it mean that most of this handbook doesn't address ARTIQ specifics at all?

Author: > The explicit assumption is that the intended audience already made the choice.

There's recently been discussion that some people would appreciate instructions on how to get ARTIQ working from a completely fresh operating system up. The idea was to cover that use case as well, as opposed to only people purchasing NUCs. It should be relatively trivial to strip that implication out if it's incorrect.

Author: > any operating system.

This is effectively a typo, there were meant to be more qualifiers. Embarrassing for me. Will fix it together with other suggestions (not on the correct machine at this time.)

---------------

Unless requested otherwise, ARTIQ NixOS NUCs ship with the desktop environment `GNOME <https://www.gnome.org/>`_, a popular choice across many Linux distributions. If you have used Linux before, you're likely already familiar with the interface. A set of basic desktop applications (Firefox, LibreOffice, Gimp) as well as some more specialized tools (Spyder, Jupyter, GTKWave, Wireshark) are also pre-installed and pre-configured. Later you will find that this list of installed packages is precisely specified in the ``configuration.nix`` file, and is very simple to edit to your needs.

.. tip::
    If you aren't familiar with GNOME, a double-tap of the Windows or Super key will show you the full list of available apps. Plenty of them should be recognizable. It's worth taking particular note of  "Console", which is GNOME's command line terminal emulator, in your case running the `friendly interactive shell <https://fishshell.com/>`_ or ``fish``. From this point forward, the handbook will assume you have a basic familiarity with a Linux environment and commands like ``sudo``. If not, it's probably worth finding some materials on these things before you continue.

> The hardware M-Labs ships for pre-installed ARTIQ NUCs is usually the `ASUS NUC 14RVK Pro Kit <https://www.asus.com/Websites/us/products/guxgpgyabgbrsz4o/pdf/6sa39dq3izs4int7.pdf>`_, with Intel i7-155H CPU, 32GB RAM, and a 1GB NVMe. Other options may also be available. Contact ``sales@`` for details.

Reviewer 2: I guess better not to put this into the docs, it may change any time without notice

    $ sudo nix flake update

and build a new *generation*: ::

    $ sudo nixos-rebuild boot

A reboot will be required for changes to take effect. See below to learn more about generations.

In order to update to a **new release version of ARTIQ** (or to the beta) system-wide, first change the source used in ``flake.nix``: ::

    inputs.artiq.url = git+https://github.com/m-labs/artiq.git?ref=release-<number>;

or remove ``?ref=release-<number>`` entirely for the beta branch, then update and rebuild as above. Note however that ``sudo nix flake update`` and a rebuild will be required every time to incorporate new commits, and for easy access to the beta it may be better to use a flake, as described in :doc:`installing`.

.. tip::

    The ``inputs.artiq.url`` key can also be replaced with a path to a local clone of the ARTIQ repository on your system, e.g.: ::

        path:/absolute/path/to/your/ARTIQ

    if you'd like to be able to edit the source for your system-wide installation. Again, however, for such purposes it may be better to use flakes, rather than requiring a full rebuild for each update.

Nix and declarative configuration
---------------------------------

> Before you start customizing your configuration, it may be useful to take some time to understand why NixOS functions the way it does. The main idiosyncracy and advantage which distinguishes Nix from most package managers and NixOS from most Linux distributions is a wholly *declarative* philosophy of package and environment management. In practice, what this means is that rather than simply installing packages into your system, you *define upfront* the list of packages available in your environment (or for particular, specialized environments) in a centralized configuration file.

Reviewer 2: Suggested simplification: `idiosyncracy` -> `feature`, `distinguishes Nix` -> `makes Nix different`

Since Nix goes to considerable lengths to build packages in isolation from each other, and to specify precisely the versions used in any given environment or build, this means, first of all, that environments are almost) perfectly replicable. It's always possible to recreate, for example, your user environment, or the ARTIQ development environment, in any other context, simply by copying or referencing the files which declare them. Isolation means installed packages can't 'clutter up' your system, and can always be removed or upgraded independently without affecting each other.

> Finally, reproduceability also applies to your *past* configurations. That is, as long as you save your previous configuration files, Nix allows for 'perfect' rollback. If you make changes that break your system, you can always revert to an older configuration, and return to precisely the environment you had before you made changes.

Reviewer 2: `reproduceability` -> `reproducibility`

Much as with flakes, this may take some time at a first use for large packages, but if you end up calling it repeatedly, subsequent invocations will use the Nix cache and run almost instantly. Also much as with flakes, the package is not 'installed' in a permanent sense and will disappear once the shell is closed. This is also useful in order to be able to test packages quickly before installing them into the environment.

Flakes and builds
^^^^^^^^^^^^^^^^^

Running NixOS is perfectly compatible with the other Nix features used by ARTIQ. In particular, using ``nix shell`` or ``nix develop`` with the various ARTIQ flakes, as described in :doc:`installing` and :doc:`flashing`, can continue to be very useful, especially to access different versions of ARTIQ, even simultaneously. It is also perfectly possible to install packages into your Nix profile. All of these are ways to define and access certain environments, with access to certain sets of packages, none of which will overwrite each other.

> Note that, for instance, if you have installed ARTIQ ``release-8`` into your environment, but run ``nix shell`` on the ARTIQ ``release-7`` flake, that *specific shell* will run ``release-7`` commands, whereas all others will continue to use the system default environment, with ``release-8``. You can check on this kind of behavior yourself with ``--version``.

Reviewer 3: ```suggestion
Note that, for instance, if you have installed ARTIQ ``release-8`` into your environment, but run ``nix shell`` on the ARTIQ ``release-7`` flake, that *specific shell* will run ``release-7`` commands, whereas all others will continue to use the system default environment, with ``release-8``. You can check your current artiq version with ``artiq --version``.
```

.. _nix-collect-garbage :

Garbage collection
^^^^^^^^^^^^^^^^^^

> As also noted in :doc:`installing`, Nix caches generously in ``/nix/store``, which should ensure that packages only ever have to be fetched and built once when reused repeatedly or between different environments. Old NixOS generations are also stored, and will remain available in the boot menu as long as they are stored. With time however this cache can occupy large amounts of storage space. To clear the cache, run: ::

Reviewer 3: ```suggestion
As also noted in :doc:`installing`, Nix stores everything in ``/nix/store``, the so-called ``Nix Store``. This ensures packages only ever have to be fetched and built once, even when reused repeatedly between different environments. Old NixOS generations are also stored in there, and will remain available in the boot menu upon future boots. With time however this can occupy large amounts of storage space. To garbage-collect (clean up files not used by NixOS) the Nix Store, run: ::
```

Calling this caching is misleading.
